### PR TITLE
Sema: Undo changes in chronological order in SolverTrail::undo()

### PR DESCRIPTION
**Pull request by slavapestov**
This was a loose end from my SolverScope refactoring. We would undo all fixed type variable bindings first and then undo the graph updates, because the graph updates were recorded in the wrong order with respect to type variable bindings. By refactoring the logic here a little we can undo changes in the order they were recorded, eliminating a second pass over the trail.

### CODE REVIEW

File: lib/Sema/ConstraintGraph.cpp [hunk 1]
```diff
@@ -343,9 +352,35 @@ void ConstraintGraphNode::retractFromInference(Constraint *constraint) {
   }
 }
 
-void ConstraintGraphNode::reintroduceToInference(Constraint *constraint) {
-  retractFromInference(constraint);
-  introduceToInference(constraint);
+void ConstraintGraphNode::retractFromInference(Type fixedType) {
```
Comment (xedin) [inline lib/Sema/ConstraintGraph.cpp:355]: Looks like the logic in `retractFromInference(Type)` and `introduceToInference` could be unified into `a method i.e. `updateAfterBind` that would accept a callback that either introduces or retracts constraints...

**Review by xedin — APPROVED** [review]
LGTM! I feel like dealing with constraints after binding a type variable is an overkill, we might get away with keeping the set as is but using simplifyType during BindingSet construction because all of the logic to compute flags (i.e. pontential incomplete, literal, delayed etc.) moved there.

File: lib/Sema/ConstraintGraph.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Sema/ConstraintGraph.cpp:355]: Good idea.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**